<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [68]:
import pandas as pd
import psycopg2
import requests
from bs4 import BeautifulSoup

In [109]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [37]:
# текст запроса
query_3_1 = f'''select 
                count(id)
                from public.vacancies
            '''

In [38]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [30]:
# текст запроса
query_3_2 = f'''select 
                count(id)
                from public.employers
            '''

In [31]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [32]:
# текст запроса
query_3_3 = f'''select 
                count(id)
                from public.areas
            '''

In [33]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [34]:
# текст запроса
query_3_4 = f'''select 
                count(id)
                from public.industries
            '''

In [36]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

                     выводы по предварительному анализу данных
В исследуемой базе данных представленно 49197 ваканмий от 23501 работодателя, в 1362 разных регионах, в 294 сферах деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [39]:
# текст запроса
query_4_1 = f'''select 
                ar.name area,
                count(vac.id) cnt
                from public.vacancies vac
                join public.areas ar on vac.area_id = ar.id
                group by ar.name
                order by 2 desc
            '''

In [43]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [47]:
# текст запроса
query_4_2 = f'''select 
                count(id)
                from public.vacancies
                where salary_from is not null
                or
                salary_to is not null
            '''

In [48]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [59]:
# текст запроса
query_4_3 = f'''select 
                avg(salary_to) "Значение",
                'среднее для верхней границы' "Пояснение"
                from public.vacancies
                union
                select
                avg(salary_from),
                'среднее для нижней границы'
                from public.vacancies
                order by 1 desc
            '''

In [60]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


Значение                    Пояснение
0  110536.741923  среднее для верхней границы
1   71064.657901   среднее для нижней границы

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [55]:
# текст запроса
query_4_4 = f'''select 
                vac.schedule "Тип рабочего графика",
                vac.employment " Тип трудоустройства",
                count(vac.id) "Количество вакансий"
                from public.vacancies vac 
                group by vac.schedule, vac.employment
                order by 3 desc
            '''

In [56]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


Тип рабочего графика  Тип трудоустройства  Количество вакансий
0           Полный день     Полная занятость                35367
1      Удаленная работа     Полная занятость                 7802
2         Гибкий график     Полная занятость                 1593
3      Удаленная работа  Частичная занятость                 1312
4        Сменный график     Полная занятость                  940
5           Полный день           Стажировка                  569
6        Вахтовый метод     Полная занятость                  367
7           Полный день  Частичная занятость                  347
8         Гибкий график  Частичная занятость                  312
9           Полный день     Проектная работа                  141
10     Удаленная работа     Проектная работа                  133
11        Гибкий график           Стажировка                  116
12       Сменный график  Частичная занятость                  101
13     Удаленная работа           Стажировка                   64
14        Гибкий график     Проектная работа                   18
15       Сменный график           Стажировка                   12
16       Вахтовый метод     Проектная работа                    2
17       Сменный график     Проектная работа                    1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [57]:
# текст запроса
query_4_5 = f'''select 
                vac.experience "Требуемый опыт работы",
                count(vac.id) "Количество вакансий"
                from public.vacancies vac 
                group by vac.experience
                order by 2 
            '''

In [58]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


Требуемый опыт работы  Количество вакансий
0           Более 6 лет                 1337
1             Нет опыта                 7197
2         От 3 до 6 лет                14511
3    От 1 года до 3 лет                26152

***

# выводы по детальному анализу вакансий
    Проведённый анализ показал, что:


    Лидерами по количеству вакансий в исследуемой БД - Крупнейшие города России, Беларуси и Казахстана.


    у более чем половины вакансий не указана заработная плата, скорее всего она будет установленна после собеседования.


    Около 72 поцентов вакансий тип рабочего дня/трудоустройства - "Полный день / Полная занятость" на втором месте - "Удаленная работа / Полная занятость" что состовляет уже около 15 процентов


    Наиболее привлекательным для работодателей являются соискатели с опытом работы от 1 года до 3 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [61]:
# текст запроса
query_5_1 = f'''(select
                emp.name "Название компании",
                count(vac.id) "Количество вакансий"
            from public.vacancies vac
            join public.employers emp on vac.employer_id = emp.id
            group by emp.name
            order by 2 desc
            limit 1)
            union all
            (select
                emp.name "Название компании",
                count(vac.id) "Количество вакансий"
            from public.vacancies vac
            join public.employers emp on vac.employer_id = emp.id
            group by emp.name
            order by 2 desc
            offset 4
            limit 1)
            '''

In [82]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


Название компании  Количество вакансий
0            Яндекс                 1933
1     Газпром нефть                  331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [62]:
# текст запроса
query_5_2 = f'''(select
                ar.name "Название региона",
                count(vac.id) "Количество вакансий",
                count(emp.id) "Количество Работодателей"
            from public.areas ar
            left join public.vacancies vac on vac.area_id = ar.id
            join public.employers emp on emp.area = ar.id
            group by ar.name
            having count(vac.id) = 0
            order by count(vac.id), count(emp.id) desc
            limit 1)
            union all
            (select
                ar.name "Название региона",
                count(vac.id) "Количество вакансий",
                count(emp.id) "Количество Работодателей"
            from public.areas ar
            left join public.vacancies vac on vac.area_id = ar.id
            join public.employers emp on emp.area = ar.id
            group by ar.name
            having count(vac.id) > 0
            order by count(vac.id) desc, count(emp.id))
            '''

In [81]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


Название региона  Количество вакансий  Количество Работодателей
0             Россия                    0                       410
1             Москва             31272712                  31272712
2    Санкт-Петербург              6320667                   6320667
3              Минск              2354880                   2354880
4             Алматы              1364132                   1364132
..               ...                  ...                       ...
464           Баймак                    1                         1
465         Тихорецк                    1                         1
466            Калач                    1                         1
467        Далматово                    1                         1
468            Себеж                    1                         1

[469 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [63]:
# текст запроса
query_5_3 = f'''select
            emp.name "Название работодателя",
            count(distinct vac.area_id) "Количество Регионов"
            from public.vacancies vac 
            join public.employers emp on emp.id = vac.employer_id
            group by emp.name
            order by count(distinct vac.area_id) desc 
            '''

In [80]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


Название работодателя  Количество Регионов
0                      Яндекс                  181
1                  Ростелеком                  152
2                  Спецремонт                  116
3      Поляков Денис Иванович                   88
4                   ООО ЕФИН                    71
...                       ...                  ...
14761                  UniSol                    1
14762            UNISTORY LLC                    1
14763                   UNIT6                    1
14764     United Distribution                    1
14765              UNITEL LLC                    1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [83]:
# текст запроса
query_5_4 = f'''select
            count(emp.id) - count(ei.employer_id) "Количество работодателей"
            from
            public.employers emp
            left join public.employers_industries ei on emp.id = ei.employer_id
            '''

In [84]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


Количество работодателей
0                      8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [66]:
# текст запроса
query_5_5 = f'''select
            emp.name "Название работодателя",
            count(ei.industry_id) "Количество сфер деятельности"
            from
            public.employers emp
            left join public.employers_industries ei on emp.id = ei.employer_id
            group by emp.name
            having count(ei.industry_id) = 4
            order by 1
            offset 2
            limit 1
            '''

In [78]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


Название работодателя  Количество сфер деятельности
0                  2ГИС                             4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [85]:
# текст запроса
query_5_6 = f'''select
            count(ei.industry_id) "Количество работодателей"
            from
            public.employers emp
            left join public.employers_industries ei on emp.id = ei.employer_id
            join public.industries ind on ind.id = ei.industry_id
            group by ind.name
            having ind.name = 'Разработка программного обеспечения' 
            '''

In [86]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


Количество работодателей
0                      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [70]:
# код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Задаём адрес ресурса
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
# локализуем таблицу с городами по указанному классу
page_table = page.find("table", {'class': 'standard sortable'})
# сохраняем в переменную все тэги типа 'а' 
links = page_table.find_all('a')
mil_lst = [link.text for link in links[4:]]

mil_cities = tuple(mil_lst)

In [75]:
# текст запроса
query_5_7 = f'''select
            ar.name "Город",
            count(vac.id) "Вакансии от Яндекс"
            from public.areas ar
            join public.vacancies vac on ar.id = vac.area_id
            join public.employers emp on emp.id = vac.employer_id
            where ar.name in {mil_cities} 
            and emp.name = 'Яндекс'
            group by ar.name
            union all
            select
            'Total',
            count(vac.id) "Вакансии от Яндекс"
            from public.areas ar
            join public.vacancies vac on ar.id = vac.area_id
            join public.employers emp on emp.id = vac.employer_id
            where ar.name in {mil_cities}
            and emp.name = 'Яндекс'
            order by 2
            '''

In [76]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


Город  Вакансии от Яндекс
0              Омск                  21
1         Челябинск                  22
2        Красноярск                  23
3         Волгоград                  24
4             Пермь                  25
5            Казань                  25
6    Ростов-на-Дону                  25
7               Уфа                  26
8            Самара                  26
9         Краснодар                  30
10          Воронеж                  32
11      Новосибирск                  35
12  Нижний Новгород                  36
13     Екатеринбург                  39
14  Санкт-Петербург                  42
15           Москва                  54
16            Total                 485

***

# выводы по анализу работодателей
    Болше всего вакансий предоставляют компании рботающие в сфере IT и банковских услуг.

    Многие компании предостовляют вакансии в нескольких регионах, что говорит о наличии у них крупной сети филиалов.

    В некоторых компаниях указано несколько сфер деятельности, но у 35% компаний сфера деятельности не указана вовсе, возможно следует сделать этот пунт обязательным к заполнениюю

    У около 15% компаний в сфере деятельности указано Разработка ПО, что говорит о большой востребованности этой сферы.

    Компания "Яндекс" разместила от 20 до 54 вакансий во всех городах-миллиониках РФ  
    

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [99]:
# текст запроса
query_6_1 = f'''select
            count(id) "Вакансии связанных с датой"
            from public.vacancies
            where name ilike '%data%'
                OR name ilike '%данн%' 
            '''

In [100]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


Вакансии связанных с датой
0                        1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [97]:
# текст запроса
query_6_2 = f'''select
            count(*) "Вакансии для начинающего ДС"
            from public.vacancies
            where (name ilike '%data scientist%'
                or name ilike '%data science%'
                or name ilike '%исследователь данных%'
                or (name like '%ML%' and name not like'%HTML%')
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                and (name ilike '%junior%'
                or experience = 'Нет опыта'
                or employment = 'Стажировка') 
            '''

In [98]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


Вакансии для начинающего ДС
0                           51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [101]:
# текст запроса
query_6_3 = f'''select
            count(*) "Вакансии ДС где важны знания SQL"
            from public.vacancies
            where (name ilike '%data scientist%'
                or name ilike '%data science%'
                or name ilike '%исследователь данных%'
                or (name like '%ML%' and name not like'%HTML%')
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                and (key_skills ilike '%SQL%' or key_skills ilike '%postgres%') 
            '''

In [102]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


Вакансии ДС где важны знания SQL
0                               201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [103]:
# текст запроса
query_6_4 = f'''select
            count(*) "Вакансии ДС где важны знания Python"
            from public.vacancies
            where (name ilike '%data scientist%'
                or name ilike '%data science%'
                or name ilike '%исследователь данных%'
                or (name like '%ML%' and name not like'%HTML%')
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                and key_skills ilike '%Python%' 
            '''

In [104]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


Вакансии ДС где важны знания Python
0                                  351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [105]:
# текст запроса
query_6_5 = f'''select
            round(
                avg(
                    1+ length(key_skills) - length(replace(key_skills, CHR(9), ''))
                ),
            2) "Среднее количество ключевых навыков"
            from public.vacancies
            where (name ilike '%data scientist%'
                or name ilike '%data science%'
                or name ilike '%исследователь данных%'
                or (name like '%ML%' and name not like'%HTML%')
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                and key_skills is not null
            '''
            #
            #  1 + length(key_skills) - length(replace(key_skills, CHR(9), ''))
            # В этой строчке находим количество символов переноса строки путём нахождения разницы количества символов в ячейке с ними и без них.

In [106]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


Среднее количество ключевых навык
0                               6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [107]:
# текст запроса
query_6_6 = f'''select
            experience "Опыт работы",
            round(
                avg(
                    coalesce(
                        (salary_from + salary_to)/2, salary_to, salary_from
                    )
                )
            ) "Средняя зарплата"
            from public.vacancies
            where (name ilike '%data scientist%'
                or name ilike '%data science%'
                or name ilike '%исследователь данных%'
                or (name like '%ML%' and name not like'%HTML%')
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                and (salary_to is not null or salary_from is not null)
                group by experience
            '''

In [108]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


Опыт работы  Средняя зарплата
0           Нет опыта           74643.0
1  От 1 года до 3 лет          139675.0
2       От 3 до 6 лет          243115.0

***

# выводы по предметному анализу
    Почти четверть вакансий связаных с датой относятся к data science

    Для начинающего дата саентиста в исследуемой БД можно найти 51 вакансию, что состовляет примерно 10% от общего числа таких ваккансий, не очень болшой процент возможно связан с тем, что работодатель сильнее заинтересован в приёме на работу более опытного специалиста, на которого не прийдётся затрачивать дополнительное время на обучение.

    Требования знаний SQL postgres и Python указаны в большинстве вакансий, что подчёркивает важность наличия этих навыков у соискателей для эфективной работы.

    В среднем работодатели требуют владения более чем шестью навыками, это может означать что работодатель заинтересован в мультизадачности своих сотрудников.

# Общий вывод по проекту

# подведем итог исследования
В исследуемой БД представлены данные о 49197 вакансиях от 23501 работодателя. Наибольшее количество вакансий представленно в крупнейших с точки зрения количества населения горадах. 

В более чем половине вакансий не указана заробоотная плата. Скорее всего она будет установленна по результатам собеседования в зависимости от знаний и навыков конкретного кандидата.

Наиболее привлекательными для работодателей являются соискатели с опытом работы от года до трёх лет, это может быть связано с тем, что такой специалист уже имеет достаточно навыков для самостоятельной работы и при этом не требует слишком высокой оплаты труда.

Больше всего компаний в исследуемой БД указало сверой деятельности IT Это указывает на то что эта сфера является очень востребованной на рынке и перспективной в плане труддоустройства.

Из всех вакансий связаных с датой примерно четверть это вакансии для дата саентиста и даже начинающий специалист без опыта работы в этой сфере сможет найти вакансию для себя, главное быть разностаронне развитым в плане профессиональных навыков. 

In [ ]:

# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

# 7.Определим какие сферы деятельности чаще всего указывают работодатели

In [112]:
query_7 = f'''select 
            ind.name "Сфера деятельность",
            count(vac.id) "Количество"
            from public.industries ind
            join public.employers_industries ei on ind.id = ei.industry_id
            join public.employers emp on emp.id = ei.employer_id
            join public.vacancies vac on emp.id = vac.employer_id
            group by ind.name
            order by 2 desc
            '''

In [113]:
df = pd.read_sql_query(query_7, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\2767996417.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7, connection)


Сфера деятельность  Количество
0                  Разработка программного обеспечения       12499
1    Системная интеграция,  автоматизации технологи...       11034
2    Интернет-компания (поисковики, платежные систе...        6413
3                                                 Банк        2742
4                                      Мобильная связь        1585
..                                                 ...         ...
288                               Спортивная федерация           5
289                                          Химчистка           4
290             Ассоциация в сфере культуры, искусства           2
291                                          Автошкола           1
292                                  Ритуальные услуги           1

[293 rows x 2 columns]

# 8.Определим ТОП-5 популярных вакансий

In [114]:
query_8 = f'''select 
            name "Название вакансии",
            count(id) "Количество"
            from public.vacancies
            group by name
            order by 2 desc
            limit 5
            '''

In [115]:
df = pd.read_sql_query(query_8, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\1280661075.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_8, connection)


Название вакансии  Количество
0           Системный администратор        1761
1                    Программист 1С        1153
2                   Бизнес-аналитик         488
3  Специалист технической поддержки         457
4                Системный аналитик         432

# 9. Определим максимальную и минимальную ЗП для дата саентиста на удалённой работе

In [116]:
query_9 = f'''select 
            experience "Опыт работы",
            round(min(coalesce((salary_from + salary_to)/2, salary_to, salary_from))) "Минимальная зарплата",
            round(max(coalesce((salary_from + salary_to)/2, salary_to, salary_from))) "Максимальная зарплата"                    
            from public.vacancies
            where (name ilike '%data scientist%'
                or name ilike '%data science%'
                or name ilike '%исследователь данных%'
                or (name like '%ML%' and name not like'%HTML%')
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                and (salary_to is not null or salary_from is not null)
                and schedule = 'Удаленная работа'
            group by experience
            '''

In [117]:
df = pd.read_sql_query(query_9, connection)
df

C:\Users\Ghost\AppData\Local\Temp\ipykernel_6716\722870376.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_9, connection)


Опыт работы  Минимальная зарплата  Максимальная зарплата
0  От 1 года до 3 лет               75000.0               275000.0
1       От 3 до 6 лет              119000.0               300000.0